File: src/app/components/projects/projectTaskAction.ts
import type {
  TaskExecutionTelemetryResponse,
  TaskGraphInsightNode,
  TodoResponse,
} from '../../types/api';

export function projectTaskAction(
  task: TodoResponse,
  insight: TaskGraphInsightNode | undefined,
  telemetry: TaskExecutionTelemetryResponse | undefined,
  telemetryAvailable: boolean,
): { kind: 'run' | 'thread' | 'details'; label: string; runId?: string } {
  const status = telemetry?.latest_run_status;
  if (
    status &&
    ['queued', 'starting', 'running', 'waiting_input', 'waiting_review'].includes(status)
  ) {
    if (!telemetry?.latest_run_id) return { kind: 'details', label: 'Details' };
    return {
      kind: 'thread',
      runId: telemetry.latest_run_id,
      label:
        status === 'waiting_review'
          ? 'Review result'
          : status === 'waiting_input'
            ? 'Needs input'
            : 'Open run',
    };
  }
  if (
    telemetryAvailable &&
    task.status === 'pending' &&
    insight?.is_ready &&
    !insight.is_container
  ) {
    return { kind: 'run', label: 'Run agent' };
  }
  return { kind: 'details', label: 'Details' };
}

File: android/feature/tasks/src/main/java/com/clawchat/android/feature/tasks/ProjectTaskAction.kt
package com.clawchat.android.feature.tasks

import com.clawchat.android.core.data.model.AgentRunStatus
import com.clawchat.android.core.data.model.ProjectNode
import com.clawchat.android.core.data.model.ProjectTaskRun

internal enum class ProjectTaskAction { RUN, PROGRESS, INPUT, REVIEW, DETAILS }

internal fun projectTaskAction(node: ProjectNode, run: ProjectTaskRun?, available: Boolean): ProjectTaskAction {
    val action = when (run?.status) {
        AgentRunStatus.QUEUED, AgentRunStatus.STARTING, AgentRunStatus.RUNNING -> ProjectTaskAction.PROGRESS
        AgentRunStatus.WAITING_INPUT -> ProjectTaskAction.INPUT
        AgentRunStatus.WAITING_REVIEW -> ProjectTaskAction.REVIEW
        else -> null
    }
    if (action != null) return if (run?.runId != null) action else ProjectTaskAction.DETAILS
    return if (available && node.isReady && !node.isContainer && node.executionState == "ready") ProjectTaskAction.RUN
        else ProjectTaskAction.DETAILS
}
